docs/dev/query/msq/tree-lifecycle.md

```diff
@@ -0,0 +1,99 @@
+the query plan is not optimized yet..
+
+## Optimization
+
+If Apache Pinot were following the Calcite architecture, this phase would optimize the logical `RelNode`s into
+physical `RelNode`s.
+But Pinot just partially follows this model.
+During optimization apply the rules defined in `PinotQueryRuleSets`, which transform the `RelNode`s in several ways:
+1. Apply some common relational algebra rules. For example, push down filters, prune project columns that are not used,
+etc.
+2. Apply some Pinot specific rules. For example, tries to optimize semi-joins into pipeline breakers, add exchanges
+to separate the query into multiple stages, etc.
+
+But these nodes are still not executable.
+
+## PlanNodes and DispatchableSubPlan
+
+In this phase, still in `QueryExecutor`, the `RelNode`s are converted to `PlanNode`s using `PinotLogicalQueryPlanner`.
+
+These `PlanNode`s were originally created to be able to serialize the query plan and send it to the servers, but they
+ended up being used to apply some final optimizations and are actually used to calculate how to distribute the
+stages between servers.
+Specifically in this phase we decide how many workers will execute each stage.
+These workers are distributed between the servers, so a server may execute the same stage more than once.
+
+## Serialization and deserialization
+The `PlanNode`s are sent over GRPC.
+The definition of the RPC methods is in `pinot-common/src/main/proto/plan.proto` and the protobuf used to serialize
+the plans is in `pinot-common/src/main/proto/plan.proto`.
+
+## Multi-stage operators
+
+The plan in protobuf format is received by the server and deserialized into `Worker.StagePlan`, generated by protobuf.
+These objects are then transformed back into `PlanNode`s using `QueryPlanSerDeUtils` and then transformed into
+`MultiStageOperator`s by MSQ `QueryRunner` (do not confuse with `QueryRunner` in SSQ).
+
+During this phase, two interesting things happen:
+1. The mailboxes are created. These are the objects that will be used to send the data to workers executing the stages
+that depend on the output of other stages. Therefore for each worker, send mailboxes are created to send the result
+of the stage and receive mailboxes are created to receive the input of the stage.
+2. In the leaf stages, the `PlanNode` is analyzed to look for all the work that could be done by SSQ.
+This is almost any table read, filter, aggregate and project whose children are also supported by SSQ.
+For example joins and window functions are not supported by SSQ.
+See `PlanNodeToOpChain` to learn more about the SSQ boundary.
+This subplan is then transformed into a single `LeafStageTransferableBlockOperator`, which uses SSQ to execute that
+part of the query. This is done mainly in order to do not have to rewrite all the optimizations we have in SSQ.
+
+Once the `MultiStageOperator`s are created, they can be [executed](execution.md) or explained.
```

Comment:
A few visual examples on how a query goes from its SQL string representation to the final MSE Operators hierarchy, going through syntax tree, relational tree, dispatchable plan, etc would be awesome. Costly and probably a bit verbose, but very useful for newcomers.